src/arns.lua

```diff
@@ -977,32 +976,27 @@ function arns.pruneRecords(currentTimestamp, lastGracePeriodEntryEndTimestamp)
 		return prunedRecords, newGracePeriodRecords
 	end
 
-	--- @type Timestamp|nil
-	local minNextEndTimestamp
+	-- identify any records that are leases and that have expired, account for a two week grace period in seconds
+	NextRecordsPruneTimestamp = nil
 
-	-- identify any records that are leases and that have expired, account for a one week grace period in seconds
-	for name, record in pairs(arns.getRecords()) do
+	-- note: use unsafe to avoid copying all the records, but be careful not to modify the records directly here
+	for name, record in pairs(arns.getRecordsUnsafe()) do
 		if arns.recordExpired(record, currentTimestamp) then
-			prunedRecords[name] = record
-			NameRegistry.records[name] = nil
+			print("Pruning record " .. name .. " because it has expired")
+			prunedRecords[name] = arns.removeRecord(name)
 		elseif arns.recordInGracePeriod(record, currentTimestamp) then
 			if record.endTimestamp > lastGracePeriodEntryEndTimestamp then
+				print(
+					"Adding record " .. name .. " to new grace period records because it has entered it's grace period"
+				)
 				newGracePeriodRecords[name] = record
 			end
 			-- Make sure we prune when the grace period is over
 			arns.scheduleNextRecordsPrune(record.endTimestamp + constants.gracePeriodMs)
 		elseif record.endTimestamp then
-			-- find the next prune timestamp
-			--- @diagnostic disable-next-line: param-type-mismatch
-			minNextEndTimestamp = math.min(minNextEndTimestamp or record.endTimestamp, record.endTimestamp)
+			arns.scheduleNextRecordsPrune(record.endTimestamp)
 		end
 	end
-
-	-- Reset the next pruning timestamp now that pruning has completed
-	NextRecordsPruneTimestamp = nil
-	if minNextEndTimestamp then
-		arns.scheduleNextRecordsPrune(minNextEndTimestamp)
-	end
```

Comment:
this is the cause of the bug. `minNextEndTimestamp` is only set/updated for records that are NOT NOR IN THEIR GRACE PERIOD (if statement above). Because `NextRecordsPruneTimestamp` is set to `nil` above, this value will ALWAYS be updated to `minNextEndTimestamp`  when calling `arns.scheduleNextRecordsPrune`. Ultimately IGNORING records the minimum that SHOULD be set by records entering their grace period.